zebROS_ws/src/teleop_joystick_control/src/2023_snap_to_nearest_conecube.cpp

```diff
@@ -0,0 +1,176 @@
+        ros::Subscriber object_detection_sub_;
+        ros::Subscriber cmd_vel_sub_;
+        ros::Publisher nearest_cone_pub_;
+        ros::Publisher nearest_cube_pub_;
+        double nearest_cargo_angle_;
+        double nearest_opponent_cargo_angle_;
+        tf2_ros::Buffer tf_buffer_;
+        tf2_ros::TransformListener tf_listener_;
+        geometry_msgs::TwistStamped cmd_vel_out_;
+        ros::Subscriber imu_sub_;
+        boost::circular_buffer<std::pair<ros::Time, double>> imu_cb_;
+        std::mutex buffer_mutex_;
+        ddynamic_reconfigure::DDynamicReconfigure ddr_;
+        double zed_time_offset_;
+
+
+    public:
+
+        SnapToConeCube2023() : tf_listener_(tf_buffer_), imu_cb_(200), zed_time_offset_(0.01)
+        {
+
+        }
+
+        void init()
+        {
+            nearest_cone_pub_ = nh_.advertise<std_msgs::Float64>("nearest_cone_angle", 1);
+            nearest_cube_pub_ = nh_.advertise<std_msgs::Float64>("nearest_cube_angle", 1);
+            object_detection_sub_ = nh_.subscribe("/tf_object_detection/object_detection_world_filtered", 1, &SnapToConeCube2023::objectDetectionCallback, this);
+            cmd_vel_sub_ = nh_.subscribe("/frcrobot_jetson/swerve_drive_controller/cmd_vel_out", 1, &SnapToConeCube2023::cmdVelSub, this);
+            imu_sub_ = nh_.subscribe("/imu/zeroed_imu", 1, &SnapToConeCube2023::imuCallback, this);
+            ROS_INFO_STREAM("snap_to_nearest_conecube_2023 : initialized");
+            ddr_.registerVariable<double>("zed_time_offset", &zed_time_offset_, "", 0, 1);
+            ddr_.publishServicesTopics();
+        }
+
+        void imuCallback(const sensor_msgs::Imu &imuState)
+        {
+            const tf2::Quaternion imuQuat(imuState.orientation.x, imuState.orientation.y, imuState.orientation.z, imuState.orientation.w);
+            double roll;
+            double pitch;
+            double yaw;
+            tf2::Matrix3x3(imuQuat).getRPY(roll, pitch, yaw);
+
+            if (std::isfinite(yaw)) // ignore NaN results
+            {
+                std::unique_lock<std::mutex> lock_that_is_unique(buffer_mutex_);
+                imu_cb_.push_back({imuState.header.stamp, -yaw});
+            }
+        }
+
+        void cmdVelSub(const geometry_msgs::TwistStamped &msg) {
+            cmd_vel_out_ = msg;
+        }
+
+        void objectDetectionCallback(const field_obj::Detection &msg)
+        {
+            double imu_angle = 0;
+            if (imu_cb_.size() != 0) {
+                std::unique_lock<std::mutex> lock_that_is_unique(buffer_mutex_);
+                for (boost::circular_buffer<std::pair<ros::Time, double>>::reverse_iterator it = imu_cb_.rbegin(); it != imu_cb_.rend(); it++) {
+                    auto angle = *it;
+                    if (angle.first <= msg.header.stamp - ros::Duration(zed_time_offset_)) {
+                        imu_angle = angle.second;
+                        break;
+                    }
+                }
+            }
```

Comment:
If the imu_angle result is used, there needs to be a check that bails out if there isn't a match found in this loop. Otherwise the code will run with imu_angle = 0?